enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,197 @@
+
+### Goals
+
+The MCO will take over management of the boot image references and the stub ignition. The installer is still responsible for creating the `MachineSet` at cluster bring-up of course, but once cluster installation is complete the MCO will ensure that boot images are in sync with the latest payload. From the user standpoint, this should cause less compatibility issues as nodes will no longer need to pivot to a different version of rhcos during node scaleup.
+
+### Non-Goals
+
+- The new subcontroller does not provide a solution for UPI as it does not use `MachineSets`. We plan to support a UPI solution via documentation that is based on this workflow.
+- This is meant to be a user opt-in feature, and if the user wishes to keep their boot images static it will let them do so.
+- This does not intend to solve [booting into custom pools](https://issues.redhat.com/browse/MCO-773). 
+
+## Proposal
+
+This automated flow is fairly straightforward, but will require a bit of special casing for each platform. 
+
+- The `machine-config-controller`(MCC) pod will gain a new sub-controller `machine_set_controller`(MSC) that monitors `MachineSet` changes and the `coreos-bootimages` [ConfigMap](https://github.com/openshift/installer/pull/4760).
+- Based on platform and arch type, the MSC will check if the images referenced in the `MachineSet(s)` is the same as the one in the ConfigMap. Each platform(gcp, aws...and so on) does this differently, so this is a good opportunity to split the work up between platforms and see if the implementation is effective. The ConfigMap is considered to be the golden set of bootimage values, i.e. they will never go out of date.
+- Next, it will check if the stub secret referenced is spec 3. If it is spec 2, the MSC will try create a new version of this secret by trying to translate it to spec 3. This step is platform/arch agnostic. Failure to up translate will cause a degrade and the sub-controller will exit without patching the `MachineSet`.
+- Finally, if the MSC will attempt to patch the `MachineSet` if required. Failure to do so will cause a degrade. 
+- Any other failures in the above steps will report an error; but degrades will only be in the specific cases mentioned above. Certain failures may also be as a result of an unsupported architecture or an unsupported platform. This is necessary because support for platforms will be phased in(and some platforms may not even desire this support)
+
+__Rolling back__
+
+The very first time bootimages are patched via this mechanism, the MSC will also backup the existing bootimage and secret references. This will be used to roll back the `MachineSets` which can be done by opting out of the feature. This is also an important mitigation in case things go wrong(invalid bootimage references, incorrect patching... etc).
```

Comment:
Back up how exactly?

Reply:
I can definitely flesh this out more. Our initial thoughts were a configmap that would have an entry per machineset. It would store the older boot image reference and the older stub secret name(these names will always exist in the mapi namespace as we will create a different secret if we up-translate). Anytime an opt-out event takes place, we will roll back the machineset(s) to the values in the configmap. GCing this is probably a good idea too, as machinesets get removed.

Reply:
How about something like:

- To opt-in a machineset to management, add an `io.openshift.mco-managed=true`
- MCO notices annotation, and instead of backing up the entire machineset, just stores the diff (probably just the specific values for the provider image reference and the secret reference) in another annotation `io.openshift.mco-pre-managed=<base64 encoded diff>`  (or maybe one annotation per field changed?)
- To revert the machineset, remove the `io.openshift.mco-managed=true` - MCO will revert the changes using the `io.openshift.mco-pre-managed` annotation

An admin can remove the backup annotation to clean things up when they want.

Reply:
This does make more sense - to associate them directly to a machineset object. I'll clarify the reference(I can see my current wording was vague) bit as well, it wasn't my intent to backup the whole machineset, just the reference name + secret name.  I guess we would need machine-api team's blessing before we put in annotations?

Also to clarify, the backup would be just once, correct? From the POV of the customer, I'd assume they want to go back to "factory" setting, i.e the value before any patches were done.

Reply:
This seems like a sub-topic for @JoelSpeed

Reply:
I haven't had a chance to review the whole doc yet, but, as we are working at the moment to move away from Machine API to Cluster API, anything we do here should take into account Cluster API as well.

In Cluster API, the bootimage becomes part of a `InfrastructureMachineTemplates` (eg an `AWSMachineTemplate`), and then the MachineSet references this template and creates new machines from the template.

Importantly each `InfrastructureMachineTemplate` is different per platform, and, is immutable. This will prevent an update in place style approach and would mean that the template would need to be cloned, updated during the clone, and then the MachineSet updated.

I'm wondering if there's a simpler approach employed already somewhere within CAPI, I'll do a bit of research and consult with others and see if there's any prior art for something like this.

It may be helpful to sync up on this once I've given the doc a complete read through.

Reply:
That is good to know! I'll do some more reading on Cluster API to understand how it intersects. And +1 to the sync, happy to set it up when we're ready.